### DISCUSSION TITLE
How to lock rendering to 30FPS / How to improve rendering performance

### DISCUSSION THREAD

**Discussion by hannojg** [google/filament#7803] [Q&A]
Hey,

I am looking into improving the performance of rendering on lower end devices.

So far I know what can help is:

- Enabling dynamic resolution scaling (which has its own bugs like crashing which I will report in a separate ticket shortly + it yet only works on android)
- Turning off post processing effects

Other optimisations that should be possible, but I don't understand how to do are:

- Reduce the rendering resolution. How would you accomplish that, while keeping the same view port size? 
- Render at a fixed rate like 30 FPS. I thought this way we'd cause less work on the GPU.

So I set the refresh rate:
```cpp
renderer->setDisplayInfo({ .refreshRate = 60.0f });
```

And then set for the target FPS the interval to `2`, so we have 2 / 60 = 30 FPS
```cpp
renderer->setFrameRateOptions({ .interval = 2 });
```

However that doesn't seem to change anything. I also tried setting the interval to 60, so effectively this should only render 1 frame per second, however, it keeps on rendering as fast as possible. Is this a bug or am I missing anything?

**Comment by romainguy** [comment]
> Reduce the rendering resolution. How would you accomplish that, while keeping the same view port size?

You can set the buffer size on `SurfaceView` and `TextureView` on Android (and I imagine iOS has something similar). This will reduce  the rendering resolution and the OS will take care of scaling for you. You can do this easily with `UiHelper.setDesiredSize()`. Alternatively you can set the min/max of the dynamic resolution system to be the same :D

@pixelflinger can answer the question about 30 fps.

**Reply by hannojg** [reply]
Thanks for the quick response! Will dynamic resolution scaling work also on iOS if I set min/max to the same value?

**Reply by hannojg** [reply]
Just tried setting the min/max to the same on iOS, seems to work, interesting. Doesn't really improve the performance though.
I am a bit confused by that - I am 

- rendering a model (64 Objects, 14,840 vertices, 27,091 triangles) 20 times using instances (so I am loading it just once, and then reuse 20 instances)
- Using no lightning / shadows, just IBL

and I have a performance of 10 FPS, 140% CPU utilisation on an older iPhone 7+ (certainly not the newest model, but its capable to handle other 3d content at 60 FPS) 🥶

**Reply by hannojg** [reply]
Interesting, so after disabling playing animation I got better results around ~30 FPS. So I guess there might be optimisation potential in the animation implementation 🤔

**Reply by pixelflinger** [reply]
Dynamic resolution should work on iOS and macOS with the metal backend. We have tested this successfully. 

To render at a lower frame rate, it is up to you to call the render loop at a lower rate. Those parameters you are setting on Renderer are for letting filament know what your frame rate is so that we can adjust dynamic resolution scaling if enabled. We can't limit the frame rate ourselves since in filament the application controls the rendering loop.

If you're using several instances of your objects, make sure to enable automatic instancing, which should help reduce draw overhead.
